pylon-core/src/main/kotlin/io/github/pylonmc/pylon/core/entity/display/transform/TransformBuilder.kt

```diff
@@ -0,0 +1,123 @@
+        val matrix = Matrix4f()
+        while (!components.isEmpty()) {
+            components.removeFirst().apply(matrix)
+        }
+        return matrix
+    }
+
+    /**
+     * Adjusts the transformation so that the transformation acts on the center of the block
+     * display; otherwise it would act on a corner
+     */
+    fun buildForBlockDisplay(): Matrix4f {
+        val cloned = TransformBuilder(this)
+        cloned.add(Translation(BLOCK_DISPLAY_ADJUSTMENT))
+        return build()
+    }
+
+    fun buildForItemDisplay(): Matrix4f {
+        return TransformBuilder(this).build()
+    }
+
+    fun buildForTextDisplay(): Matrix4f {
+        return TransformBuilder(this).build()
+    }
+
+    class LineBuilder {
+
+        var translation: Vector3f = Vector3f()
+        var from: Vector3f? = null
+        var to: Vector3f? = null
+        var thickness: Float? = null
+        var extraLength: Float = 0.0F
+        var roll: Float = 0.0F
+
+        fun translation(translation: Vector3f) = apply { this.translation = translation }
+        fun translation(translation: Vector3d) = translation(Vector3f(translation))
+        fun translation(x: Float, y: Float, z: Float) = translation(Vector3f(x, y, z))
+        fun translation(x: Double, y: Double, z: Double) = translation(Vector3d(x, y, z))
+        fun from(from: Vector3f) = apply { this.from = from }
+        fun from(from: Vector3d) = from(Vector3f(from))
+        fun from(x: Float, y: Float, z: Float) = from(Vector3f(x, y, z))
+        fun from(x: Double, y: Double, z: Double) = from(Vector3d(x, y, z))
+        fun to(to: Vector3f) = apply { this.to = to }
+        fun to(to: Vector3d) = to(Vector3f(to))
+        fun to(x: Float, y: Float, z: Float) = to(Vector3f(x, y, z))
+        fun to(x: Double, y: Double, z: Double) = to(Vector3d(x, y, z))
+        fun thickness(thickness: Float) = apply { this.thickness = thickness }
+        fun thickness(thickness: Double) = thickness(thickness.toFloat())
+        fun extraLength(extraLength: Float) = apply { this.extraLength = extraLength }
+        fun extraLength(extraLength: Double) = extraLength(extraLength.toFloat())
+        fun roll(roll: Float) = apply { this.roll = roll }
+        fun roll(roll: Double) = roll(roll.toFloat())
+
+        fun build(): TransformBuilder {
+            if ((from == null) or (to == null) or (thickness == null)) {
+                throw IllegalStateException("From, to, and thickness in LineBuilder must all be specified")
+            }
+            val midpoint = TransformUtil.getMidpoint(from!!, to!!).add(translation)
+            return TransformBuilder()
+                .translate(midpoint)
+                .lookAlong(from!!, to!!)
+                .rotate(0.0F, 0.0F, roll)
+                .scale(Vector3f(thickness!!, thickness!!, from!!.distance(to!!) + extraLength))
+        }
+    }
+
+    companion object {
+        private val BLOCK_DISPLAY_ADJUSTMENT: Vector3f = Vector3f(-0.5F)
```

Comment:
```suggestion
        @JvmField
        val BLOCK_DISPLAY_ADJUSTMENT: Vector3f = Vector3f(-0.5F, -0.5F, -0.5F)
```
other plugins might end up using this, no reason to hide it from them, and I think defining all 3 params explicitly makes it a bit clearer / more readable